src/engine/sched.c

```diff
@@ -2077,6 +2101,71 @@ sched_try_relax(struct dss_xstream *dx, ABT_pool *pools, uint32_t running)
 	d_tm_inc_counter(info->si_stats.ss_relax_time, sleep_time);
 }
 
+/* Use SWIM xstream to monitor scheduler activities of SYS & VOS xstreams */
+static void
+sched_xs_monitor(struct dss_xstream *cur_dx)
+{
+	struct dss_xstream    *dx;
+	struct sched_info     *info, *cur_info;
+	struct sched_hist_seq *hist;
+	unsigned int           gap;
+	int                    rc, i, inactive_tgt, inactive_id = -1;
+
+	D_ASSERT(is_monitor_xs(cur_dx));
+	/* Monitor isn't enabled */
+	if (sched_inactive_max == 0)
+		return;
+
+	cur_info = &cur_dx->dx_sched_info;
+	/* Check schedule activities every two seconds */
+	hist = &cur_info->si_hist_seqs[cur_dx->dx_xs_id];
+	if ((hist->sm_last_ts + 2000) > cur_info->si_cur_ts)
+		return;
+
+	hist->sm_last_ts = cur_info->si_cur_ts;
+	if (!dss_sched_monitor_enter())
+		return;
+
+	/* Accessing other xstream data without locking */
```

Comment:
[Nit] This comment makes it sounds like we are introducing data races, but after checking every access to `xstream_data`, I realize that you actually mean "these reads never race with the writes", thanks to `dss_sched_monitor_enter` and the "open barrier". Is that right?

Reply:
Hmm, wait... Isn't the read from `si_cur_ts` racing with writes to the field from other xstream?

Reply:
Maybe the proper solution is to change `si_cur_ts` and `si_cur_seq` to `atomic_ullong`, and access them with `atomic_load`/`atomic_store`?